### DISCUSSION TITLE
How can I boot Avalonia Application at startup in Linux DRM?

### DISCUSSION THREAD

**Discussion by kyuranger** [AvaloniaUI/Avalonia#15651] [Q&A]
I want my Avalonia Application to run when the raspberry pi (raspbian lite) startup.

The application runs well when I input this in terminal:
sudo ./test/test --drm

Then I try the way by editing rc.local and add the code:
sudo ./test/test --drm
above the line of exit 0

When I reboot the raspberry pi, it reports an error:
sudo: ./test/test:command not found

What's wrong with my code?

Although there are many ways to set an application startup in raspberry. Whereas, most of them need to add useless services or install python framework. I want to use the simplest way to achieve this without loading unnecessary frameworks or services (which will cause the Raspberry Pi to run slower).

**Comment by thevortexcloud** [comment]
`./` refers to the **current** directory. Which is very likely not where your init system is running from (if you get it to output the result of `pwd` it will tell you what `./` currently refers to). Try using an absolute path. That said if your OS supports Systemd (fairly likely) you can also make a systemd service for this which will likely work a lot better. You can also use systemd to auto restart your app if it crashes if you register it as a proper service. Which is likely something you want to do anyway if you are making an emedded appliance sort of thing.

On a related note: .NET actually has official systemd integration. Although you probably don't need it for this.

https://devblogs.microsoft.com/dotnet/net-core-and-systemd/

**Reply by kyuranger** [reply]
I tried your way by using the absolute path, now raspberry reports another error:
```
rc,local[569]: Unhandled exception, System, InvalidOperationException: Cannot read keys when either application does not have a console or when console input has been redirected, Try Console.Read.
rc,local[569]:at System,ConsolePal,ReadKey(Boolean intercept)
rc.local[569]:at test,Program,<>c,<SilenceConsole>b_ 1 gc) in D:\Project\test\test\Program.cs:line 41
rc.local[569]:Aboarted
```

It seems now the program started but somethin wrong with .Net

The line 41 in Program.cs is in
```
 private static void SilenceConsole()
 {
     new Thread(() =>
     {
         Console.CursorVisible = false;
         while (true)
             Console.ReadKey(true);
     })
     { IsBackground = true }.Start();
 }
```

 Console.ReadKey(true); is just the line 41

Then I tried to run the command sudo ./test/test --drm in terminal again, but it works well. I don't quite understand what the problem is?

**Reply by thevortexcloud** [reply]
My guess is you are not running as the correct user/in a real session/the terminal is not actually ready yet/some other required device has not been created yet. However I am not an expert on what Linux does at the init stage and how .NET's console APIs work on Linux, so I can only speculate what is happening here.

From the sound of things you would likely be better off using systemd, as it's much easier to control this sort of thing (EG you can make it so it only starts after certain other services are started and also run it as a specific user). You are likely already using it for many other things anyway without realising it. It's included and enabled by default in most modern distros.

**Reply by kyuranger** [reply]
I tried using the systemd to solve this but systemd reports the same error also.

**Reply by kyuranger** [reply]
BTW, I think this should be caused by Avalonia (I guess some tricks may need to be added, such as delayed startup).

**Comment by kyuranger** [comment]
I tried to use rc.local to boot with absolute path and modify the code in program.cs
```
 private static void SilenceConsole()
 {
     new Thread(() =>
     {
         Console.CursorVisible = false;
         while (true)
             Console.ReadKey(true);
     })
     { IsBackground = true }.Start();
 }
```

Change this line:
`Console.ReadKey(true);`
to
`Console.Read();`

Now although it works. But I don't know what the difference is between two codes above. And I'm not sure if it will cause other problems because I change this line of code
